src/input/joystick.c

```diff
@@ -0,0 +1,720 @@
+                        break;
+                    case JOYSTICK_HAT_DOWN:
+                        current_value = hat->bottom;
+                        break;
+                    case JOYSTICK_HAT_RIGHT:
+                        current_value = hat->right;
+                        break;
+                    default:
+                        current_value = 0;
+                        log_info("Invalid hat value for hat", 0, element_id);
+                        break;
+                }
+                break;
+            }
+            default:
+                continue;
+        }
+        if (current_value == 0) {
+            input->element = JOYSTICK_ELEMENT_NONE;
+            input->value = 0;
+            return 0;
+        }
+        // Element precedence. When two joystick elements are required for the input
+        // to be registered, the value is obtained from the highest priority element.
+        // Element priority in descending order: axis, trackball, button, hat
+        if (input->element == JOYSTICK_ELEMENT_NONE || current_element < input->element) {
+            input->element = current_element;
+            input->value = current_value;
+        }
+    }
+    return input->value != 0;
+}
+
+static void set_input_state(mapped_input *input)
+{
+    if (input->value) {
+        if (input->state == INPUT_STATE_IS_UP || input->state == INPUT_STATE_WENT_UP) {
+            input->state = INPUT_STATE_WENT_DOWN;
+        } else {
+            input->state = INPUT_STATE_IS_DOWN;
+        }
+    } else {
+        if (input->state == INPUT_STATE_IS_DOWN || input->state == INPUT_STATE_WENT_DOWN) {
+            input->state = INPUT_STATE_WENT_UP;
+        } else {
+            input->state = INPUT_STATE_IS_UP;
+        }
+    }
+}
+
+static int get_joystick_input_for_action(mapping_action action, mapped_input *input)
+{
+    static mapped_input dummy_input;
+    if (!input) {
+        input = &dummy_input;
+    }
+
+    input->value = 0;
+    input->element = JOYSTICK_ELEMENT_NONE;
+
+    for (int i = 0; i < MAX_CONTROLLERS; ++i) {
+        const joystick_info *joystick = &data.joystick[i];
+        if (!joystick->connected) {
+            continue;
+        }
+        const joystick_model *model = joystick->model;
+        for (int j = 0; j < model->num_mappings; j++) {
+            if (model->mapping[j].action == action && get_input_for_mapping(joystick, &model->mapping[j], input)) {
+                set_input_state(input);
+                return 1;
+            }
+        }
+    }
+    set_input_state(input);
+    return 0;
+}
+
+static void translate_input_for_element(mapped_input *input, joystick_element translated_element)
+{
+    if (input->element == translated_element) {
+        return;
+    }
+    if (translated_element == JOYSTICK_ELEMENT_AXIS) {
+        switch(input->element)
+        {
+            case JOYSTICK_ELEMENT_TRACKBALL:
+                input->value *= 5; // TODO is this a high enough value? I doubt it
```

Comment:
Do we need to resolve this TODO before merging?

Reply:
No, at least not until sometime has an actual trackball to test!

Reply:
Hmm, ok. Can we at least move the `5` to a constant and add a comment there instead? I'm hesitant to merge code with TODOs in them.

Reply:
Sure!

If it helps your OCD, remember this is just phase one - we can't test trackballs anyway, because there's no mapping support for custom controllers!

And we'll need someone with an actual trackball to see if the ratio of `5` is under or oversensitive.